src/key-value-engine/B-Tree-vs-Log-Structured-Merge-Tree.md

```diff
@@ -0,0 +1,130 @@
+
+## Metrics
+
+In general, there are three most important metrics to measure the performance of a data structure, which include write amplification, read amplification, and space amplification. This section aims to describe these metrics. 
+
+### Write Amplification
+
+`Write amplification` is the ratio of the amount of data written to the storage device versus the amount of data written to the database. 
+
+For example, if you are writing 10 MB to the database and you observe 30 MB disk write rate, your write amplification is 3.
+
+Flash memory and solid-state drives (SSDs) can be written to only a finite number of times, so that write amplification will decrease the flash lifetime.
+
+There is another write amplification associated with the flash memory and SSDs because flash memory must be erased before it can be rewritten.
+
+### Read Amplification
+
+`Read amplification` is the number of disk reads per query. 
+
+For example, if you need to read 5 pages to answer a query, read amplification is 5. 
+
+Note that the units of write amplification and read amplification are different. Write amplification measures how much more data is written than the application thought it was writing, whereas read amplification counts the number of disk reads to perform a query.
+
+Read amplification is defined separately for point query and range queries. For range queries the range length matters (the number of rows to be fetched). 
+
+Cache is a critical factor for read amplification. For example a B-tree in the cold-cache case, a point query requires \\(O(log_BN)\\) disk reads, whereas in the warm-cache case the internal nodes of the B-tree are cached, and so a B-tree requires at most one disk read per query.
+
+### Space Amplification
+
+`Space amplification` is the ratio of the amount of data on the storage device versus the amount of data on the database. 
+
+For example, if you put 10MB in the database and this database uses 100MB on the disk, then the space amplification is 10.
+
+Generally speaking,  an data structure can optimize for at most two from read, write, and space amplification. This means one data structure is unlikely to be better than another at all three. For example a B-tree has less read amplification than an LSM-tree while an LSM-tree has less write amplification than a B-tree. 
+
+## Analysis
+
+B-tree is a generalization of [binary search tree](https://en.wikipedia.org/wiki/Binary_search_tree) in which a node can have more than two children. There are two kinds of node in B-tree, internal nodes ,and leaf nodes. A leaf node contains data records and has no children, whereas an internal node can have a variable number of child nodes within some pre-defined range, internal nodes may be joined or split. An example of B-tree appears in *Figure 1*.
```

Comment:
I think start with `B-trees` (if you're talking about all B-trees) or `The B-tree` (if you're talking about the abstract idea of the structure)